module/VuFind/src/VuFind/Http/CachingDownloader.php

```diff
@@ -0,0 +1,168 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Http
+ * @author   Mario Trojan <mario.trojan@uni-tuebingen.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+namespace VuFind\Http;
+
+use Laminas\Cache\Storage\StorageInterface;
+use VuFind\Cache\Manager as CacheManager;
+use VuFindHttp\HttpService;
+
+/**
+ * Caching downloader.
+ *
+ * @category VuFind
+ * @package  Http
+ * @author   Mario Trojan <mario.trojan@uni-tuebingen.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class CachingDownloader
+{
+    /**
+     * CacheManager to update caches if necessary.
+     *
+     * @var CacheManager
+     */
+    protected $cacheManager;
+
+    /**
+     * Stored client options for cache key generation.
+     *
+     * @var array
+     */
+    protected $cacheOptions = [];
+
+    /**
+     * Cache to use for downloads
+     *
+     * @var StorageInterface
+     */
+    protected $cache;
+
+    /**
+     * HTTP service
+     *
+     * @var HttpService
+     */
+    protected $httpService;
+
+    /**
+     * Constructor
+     *
+     * @param HttpService $httpService  HTTP service
+     * @param string      $cacheManager Base directory for cache
+     */
+    public function __construct(HttpService $httpService, CacheManager $cacheManager)
+    {
+        $this->httpService = $httpService;
+        $this->cacheManager = $cacheManager;
+        $this->setCacheId('downloader');
+    }
+
+    /**
+     * Set a different cache.
+     *
+     * @param string $cacheId Cache ID
+     *
+     * @return void
+     */
+    public function setCacheId($cacheId)
+    {
+        $cacheName = $this->cacheManager->addDownloaderCache($cacheId);
+        $this->cache = $this->cacheManager->getCache($cacheName);
+    }
+
+    /**
+     * Download a resource using the cache in the background.
+     *
+     * @param string   $url              URL
+     * @param array    $params           Request parameters (e.g. additional headers)
+     * @param callable $validateCallback Callback for validation
+     *                                   before storing to cache
+     * @param callable $decodeCallback   Callback for decoding
+     *
+     * @return mixed
+     */
+    public function download(
+        $url,
+        $params=[],
+        callable $validateCallback=null,
+        callable $decodeCallback=null
+    ) {
+        $cacheItemKey = md5($url);
+        foreach ($params as $paramKey => $paramValue) {
```

Comment:
Is there any reason not to use `http_build_query` here? Might be more concise. Also, any reason to include the parameters outside the md5 hash? I think including them in the hash would result in more consistent -- and shorter -- keys.